boards/shields/wiznet_wiz550io/wiznet_wiz550io.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2024 Grant Ramsay
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * This driver provides the initialsation sequence required for
+ * WIZ550io to configure the W5500 Ethernet controller with its
+ * embedded unique MAC address.
+ */
+
+#define DT_DRV_COMPAT wiznet_wiz550io
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(eth_w5500, CONFIG_ETHERNET_LOG_LEVEL);
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/net/ethernet.h>
+
+#include "eth.h"
+#include "eth_w5500_priv.h"
+
+struct wiz550io_config {
+	const struct device *w5500_dev;
+	struct gpio_dt_spec reset;
+	struct gpio_dt_spec ready;
+	bool use_mac_addr;
+};
+
+/* WIZ550io documentation recommends 150ms delay after HW reset
+ * for PIC12F519 MCU to configure the W5500.
+ */
+#define WIZ550IO_RESET_DELAY K_MSEC(150)
+
+static int wiz550io_init(const struct wiz550io_config *config)
+{
+	int err;
+	const struct w5500_config *w5500_config = config->w5500_dev->config;
+	struct w5500_runtime *w5500_runtime = config->w5500_dev->data;
+
+	if (!spi_is_ready_dt(&w5500_config->spi)) {
+		LOG_ERR("SPI master port %s not ready", w5500_config->spi.bus->name);
+		return -EINVAL;
+	}
+
+	if (!gpio_is_ready_dt(&config->reset)) {
+		LOG_ERR("GPIO port %s not ready", config->reset.port->name);
+		return -EINVAL;
+	}
+
+	if (gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT)) {
+		LOG_ERR("Unable to configure GPIO pin %u", config->reset.pin);
+		return -EINVAL;
+	}
+
+	/* Hold reset pin low for 500us to perform a hardware reset */
+	gpio_pin_set_dt(&config->reset, 1);
+	k_usleep(500);
+	gpio_pin_set_dt(&config->reset, 0);
+
+	/* Wait for the device to be ready */
+	if (config->ready.port) {
+		if (!gpio_is_ready_dt(&config->ready)) {
+			LOG_ERR("GPIO port %s not ready", config->ready.port->name);
+			return -EINVAL;
+		}
+		if (gpio_pin_configure_dt(&config->ready, GPIO_INPUT)) {
+			LOG_ERR("Unable to configure GPIO pin %u", config->ready.pin);
+			return -EINVAL;
+		}
+
+		k_timepoint_t timeout = sys_timepoint_calc(WIZ550IO_RESET_DELAY);
+
+		while (!gpio_pin_get_dt(&config->ready)) {
+			if (sys_timepoint_expired(timeout)) {
+				LOG_ERR("WIZ550io not ready");
+				return -ETIMEDOUT;
+			}
+			k_msleep(1);
+		}
+	} else {
+		k_sleep(WIZ550IO_RESET_DELAY);
+	}
+
+	if (config->use_mac_addr) {
+		/* Read/assign the WIZ550io unique MAC address set by the PIC12F519
+		 * after hardware reset. It needs to be read after a hardware reset
+		 * and before software reset.
+		 */
+		err = w5500_spi_read(config->w5500_dev, W5500_SHAR,
+				     w5500_runtime->mac_addr,
```

Comment:
instead of using the driver runtime from the ethernet controller, I believe that there is an API that can be called to set the mac address https://docs.zephyrproject.org/latest/connectivity/networking/api/ethernet.html#c.ethernet_api.set_config